src/json.c

```diff
@@ -0,0 +1,1150 @@
+
+   JsonNode *port_array = json_mkarray();
+   if (tree_ports(t) > 0) {
+      for (unsigned i = 0; i < tree_ports(t); i++) {
+         json_append_element(port_array, dump_port(tree_port(t, i)));
+      }
+   }
+   json_append_member(entity_node, "port", port_array);
+   json_append_member(entity_node, "decls", dump_decls(t));
+
+   JsonNode *stmts_array = json_mkarray();
+   for (unsigned i = 0; i < tree_stmts(t); i++) {
+      json_append_element(stmts_array, dump_stmt(tree_stmt(t, i)));
+   }
+   json_append_member(entity_node, "stmts", stmts_array);
+   return entity_node;
+}
+
+static JsonNode *dump_decls(tree_t t)
+{
+   JsonNode *decls = json_mkarray();
+   const int ndecls = tree_decls(t);
+   for (unsigned i = 0; i < ndecls; i++)
+      json_append_element(decls, dump_decl(tree_decl(t, i)));
+   return decls;
+}
+
+static JsonNode *dump_arch(tree_t t)
+{
+   JsonNode *architecture_node = json_mkobject();
+   json_append_member(architecture_node, "cls", json_mkstring("architecture"));
+   json_append_member(architecture_node, "ctx", dump_context(t));
+   add_filename(architecture_node, t);
+   json_append_member(architecture_node, "name", json_mkstring(istr(tree_ident(t))));
+   json_append_member(architecture_node, "of", json_mkstring(istr(tree_ident2(t))));
+   json_append_member(architecture_node, "decls", dump_decls(t));
+   JsonNode *stmts_array = json_mkarray();
+   for (unsigned i = 0; i < tree_stmts(t); i++)
+      json_append_element(stmts_array, dump_stmt(tree_stmt(t, i)));
+   json_append_member(architecture_node, "stmts", stmts_array);
+   return architecture_node;
+}
+
+static JsonNode *dump_package(tree_t t)
+{
+   JsonNode *pkg_node = json_mkobject();
+   json_append_member(pkg_node, "cls", json_mkstring("pkg"));
+   json_append_member(pkg_node, "ctx", dump_context(t));
+   add_filename(pkg_node, t);
+   json_append_member(pkg_node, "name", json_mkstring(istr(tree_ident(t))));
+   json_append_member(pkg_node, "decls", dump_decls(t));
+   return pkg_node;
+}
+
+static JsonNode *dump_package_body(tree_t t)
+{
+   JsonNode *pkg_node = json_mkobject();
+   json_append_member(pkg_node, "cls", json_mkstring("pkg_body"));
+   json_append_member(pkg_node, "ctx", dump_context(t));
+   add_filename(pkg_node, t);
+   json_append_member(pkg_node, "name", json_mkstring(istr(tree_ident(t))));
+   json_append_member(pkg_node, "decls", dump_decls(t));
+   return pkg_node;
+}
+
+static JsonNode *dump_configuration(tree_t t)
+{
+   JsonNode *config_node = json_mkobject();
+   json_append_member(config_node, "cls", json_mkstring("config"));
+   json_append_member(config_node, "name", json_mkstring(istr(tree_ident(t))));
+   json_append_member(config_node, "of", json_mkstring(istr(tree_ident2(t))));
+   json_append_member(config_node, "decls", dump_decls(t));
+   return config_node;
+}
+
+void dump_json(tree_t *elements, unsigned int n_elements, const char *filename)
+{
+   FILE* dump_file = fopen(filename, "w");
+   if (!dump_file) {
```

Comment:
Need to print an error here if there's a problem opening the file. Maybe just call `fatal_errno(filename)`?

Reply:
Done